core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -212,6 +226,70 @@ class half_t {
   KOKKOS_FUNCTION
   half_t(unsigned long long rhs) : val(cast_to_half(rhs).val) {}
 
+  // Atomics
+  KOKKOS_FUNCTION
+  void atomic_add(const volatile half_t src) volatile {
+#ifdef __CUDA_ARCH__
+#if (CUDA_VERSION < 10000)
+    volatile float* lhs = (float*)&val;
+    volatile float* rhs = (float*)&src.val;
+    *lhs                = __half2float(val);
+    *rhs                = __half2float(src.val);
+    val                 = __float2half(lhs[0] + rhs[0]);
+#else
+    // Cuda 10 supports __half volatile stores but not volatile arithmetic
+    // operands so we use atomicAdd for cuda 10.0 instead.
+    atomicAdd(const_cast<impl_type*>(&val),
+              const_cast<const impl_type&>(src.val));
+#endif  // CUDA_VERSION
+#else
+    // 32bit arithmetic on host followed by 16bit atomic store.
+    uint16_t* val_ptr = (uint16_t*)const_cast<impl_type*>(&val);
+    const impl_type new_val =
+        __float2half(__half2float(const_cast<impl_type&>(val)) +
+                     __half2float(const_cast<const impl_type&>(src.val)));
+    __atomic_store_n(val_ptr, *((uint16_t*)&new_val), __ATOMIC_RELAXED);
```

Comment:
nope not available on all platforms.

Reply:
doesn't Kokkos::atomic_store work?

Reply:
Kokkos::atomic_store will work but I believe it uses a mutex for 2 byte scalars; shall I change to Kokkos::atomic_store ?

Reply:
ok so we need atomic_store for 2 byte types